<a href="https://colab.research.google.com/github/Sumit367/Marketing-Analytics/blob/main/Sumit_LAB8_(CNN_MNIST_dataset).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **DAB311 Introduction to Deep Learning**

### **Sumit (W0850740)**

**1. Setup**

**Install the keras package**

In [2]:
pip install tensorflow numpy mnist

In [24]:
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow import keras

# Load the dataset from Keras datasets module
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Print the shapes of the training data
print(train_images.shape)  # (60000, 28, 28)
print(train_labels.shape)  # (60000,)

(60000, 28, 28)
(60000,)


**2. Preparing the Data**

In [6]:
import numpy as np
from tensorflow.keras.datasets import mnist

# Load the dataset from Keras datasets module
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = (train_images / 255.0) - 0.5
test_images = (test_images / 255.0) - 0.5

# Reshape the images
train_images = np.expand_dims(train_images, axis=3)
test_images = np.expand_dims(test_images, axis=3)

print(train_images.shape)  # (60000, 28, 28, 1)
print(test_images.shape)   # (10000, 28, 28, 1)

(60000, 28, 28, 1)
(10000, 28, 28, 1)


**3. Building the Model**

In [7]:
from tensorflow.keras.models import Sequential

# WIP
model = Sequential([
  # layers...
])

In [8]:
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten

num_filters = 8
filter_size = 3
pool_size = 2

model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

**4. Compiling the Model**

In [9]:
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

**5. Training the Model**

In [11]:
import numpy as np
from tensorflow.keras.datasets import mnist

# Load the dataset from Keras datasets module
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Print the first label to verify
print(train_labels[0])  # This should print the first label, which might be 5

5


In [12]:
from tensorflow.keras.utils import to_categorical

model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=3,
  validation_data=(test_images, to_categorical(test_labels)),
)

Epoch 1/3
1875/1875 [==============================] - 19s 10ms/step - loss: 2.1509 - accuracy: 0.8845 - val_loss: 0.4509 - val_accuracy: 0.9353
Epoch 2/3
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3672 - accuracy: 0.9399 - val_loss: 0.3291 - val_accuracy: 0.9429
Epoch 3/3
1875/1875 [==============================] - 18s 10ms/step - loss: 0.2548 - accuracy: 0.9496 - val_loss: 0.2792 - val_accuracy: 0.9473


In [14]:
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from tensorflow.keras.utils import to_categorical

# Load the dataset from Keras datasets module
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = (train_images / 255.0) - 0.5
test_images = (test_images / 255.0) - 0.5

# Reshape the images
train_images = np.expand_dims(train_images, axis=3)
test_images = np.expand_dims(test_images, axis=3)

# Set parameters for the model
num_filters = 8
filter_size = 3
pool_size = 2

# Build the model
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compile the model
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=3,
  validation_data=(test_images, to_categorical(test_labels)),
)


Epoch 1/3
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3652 - accuracy: 0.8945 - val_loss: 0.2281 - val_accuracy: 0.9329
Epoch 2/3
1875/1875 [==============================] - 17s 9ms/step - loss: 0.1898 - accuracy: 0.9463 - val_loss: 0.1540 - val_accuracy: 0.9544
Epoch 3/3
1875/1875 [==============================] - 16s 9ms/step - loss: 0.1364 - accuracy: 0.9611 - val_loss: 0.1147 - val_accuracy: 0.9665


We achieve **95.4%** test accuracy with this simple CNN!

**6. Using the Model**

In [15]:
model.save_weights('cnn.h5')

In [16]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten

num_filters = 8
filter_size = 3
pool_size = 2

# Build the model.
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Load the model's saved weights.
model.load_weights('cnn.h5')

In [17]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

1/1 [==============================] - 0s 116ms/step
[7 2 1 0 4]
[7 2 1 0 4]


**8. Extensions**

Network Depth

In [18]:
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  Conv2D(num_filters, filter_size),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

Dropout

In [19]:
from tensorflow.keras.layers import Dropout

model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Dropout(0.5),
  Flatten(),
  Dense(10, activation='softmax'),
])

Fully-connected Layers

In [20]:
from tensorflow.keras.layers import Dense

model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

Convolution Parameters

In [21]:
# These can be changed, too!
num_filters = 8
filter_size = 3

model = Sequential([
  # See https://keras.io/layers/convolutional/#conv2d for more info.
  Conv2D(
    num_filters,
    filter_size,
    input_shape=(28, 28, 1),
    strides=2,
    padding='same',
    activation='relu',
  ),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

We achieved a test accuracy of **95.4%** with our simple initial network.